docs/development/design/build-images.rst

```diff
@@ -0,0 +1,274 @@
+Build Images
+============
+
+This document describes how Read the Docs uses the `Docker Images`_ and how they are named.
+Besides, it proposes a new way to create and name them to allow
+sharing as many image layers as possible to support more customization while keeping the stability.
+
+.. _Docker Images: https://github.com/readthedocs/readthedocs-docker-images
+
+
+Introduction
+------------
+
+We use Docker images to build user's documentation.
+Each time a build is triggered, one of our VMs picks the task
+and go through different steps:
+
+#. run some application code to spin up a Docker image into a container
+#. execute ``git`` inside the container to clone the repository
+#. analyze and parse files (``.readthedocs.yaml``) from the repository *outside* the container
+#. create the environment and install docs' dependencies inside the container
```

Comment:
I think we can add another step before this "Spin up a new Docker image based on the config file"

Reply:
@stsewd you could use the build-pattern for docker like this:
https://github.com/staticdev/enelvo-microservice/blob/master/Dockerfile

One base image does the `git clone` then this image is deleted to give birth to a cleaner one.

Reply:
Didn't know that :). But we can't use that here, since the git command used here change per project/settings, and we also execute other git operations on this container. And we don't know what build image we are going to use previous to this step.